packages/next/client/app-index.tsx

```diff
@@ -64,7 +65,11 @@ function renderReactElement(
   const reactEl = fn()
   if (!reactRoot) {
     // Unlike with createRoot, you don't need a separate root.render() call here
-    reactRoot = (ReactDOMClient as any).hydrateRoot(domEl, reactEl)
+    reactRoot = requestIdleCallback(() => {
```

Comment:
`reactRoot` here is changed from `ReactDOMClient.Root` to `IdleCallBackHandle`.

Is it intended?

Reply:
The original code works like this:

- Is variable `reactRoot` defined as a `ReactDOMClient.Root`?
  - No, `reactRoot` is null: Call `ReactDOMClient.hydrateRoot` to initiate it.
  - Yes, `reactRoot` is not null: Simply call `ReactDOMClient.Root.render` to update the existing React nodes.

That's why there is an `if (!reactRoot)` and a `reactRoot.render`. Andn now you just break it. `Returntype<typeof requestIdleCallback>` is `number`, not `ReactDOMClient.Root`.

> And one doubt, `.hydrateRoot` called only once nah?

So let's say, `reactRenderElements` itself is being called twice.

- First time `reactRenderElements` is being called, `reactRoot` is `null`.
- `reactRoot = requestIdleCallback(() => { /** */ })`, now `reactRoot` is a `number`.
- The second time `reactRenderElements` is being called, `!reactRoot` is now `false`
- `reactRenderElements` will execute `reactRoot.render`
- The code throws an error: `Uncaught TypeError: <number>.render is not a function`

Reply:
> Thanks a lot for explaining 😊
> 
> So, can we wrap `.render` with `startTransition` ? or can we do like this inside `app-next.js` ?

Or, `requestIdleCallback + startTransition` could be added inside `export function hydrate` just like this:

```ts
let hydrateIdleCallbackHandle = null | number;

export function hydrate(opts?: { onFlightCssLoaded?: () => Promise<void> }) {
  if (hydrateIdleCallbackHandle) clearIdleCallback(hydrateIdleCallbackHandle);
  hydrateIdleCallbackHandle = requestIdleCallback(() => {
    renderReactElement(appElement!, () => (
      <React.StrictMode>
        <Root>
          <RSCComponent onFlightCssLoaded={opts?.onFlightCssLoaded} />
        </Root>
      </React.StrictMode>
    ))
  });
}
```

IMHO, deferred hydration to reduce FID (First Input Delay) is a great idea, but where and how to add this is still depends on how Next.js Layout RFC is implemented.